designs/2024-config-extends/README.md

```diff
@@ -0,0 +1,613 @@
+
+```js
+
+const extender = new ConfigExtender();
+
+class FlatConfigArray {
+
+    // snip
+
+    normalize(context) {
+
+        // make sure not to make any changes if already normalized
+        if (this.isNormalized()) {
+            throw new Error("...");
+        }
+
+        // replace each element with an array
+        this.forEach((element, i) => {
+            const configs = Array.isArray(element) ? element : [element];
+            this[i] = configs.flat().map(config => extender.evaluate(config))
+        });
+
+        // proceed as usual
+        return super.normalize(context)
+            .catch(error => {
+                if (error.name === "ConfigError") {
+                    throw wrapConfigErrorWithDetails(error, this[originalLength], this[baseLength]);
+                }
+
+                throw error;
+
+            });
+    }
+}
+```
+
+## Documentation
+
+This will require documentation changes and an introductory blog post.
+
+At a minimum, these pages will have to be updated:
+
+* https://eslint.org/docs/latest/use/getting-started#configuration
+* https://eslint.org/docs/latest/use/configure/configuration-files
+* https://eslint.org/docs/latest/extend/plugins#configs-in-plugins
+* https://eslint.org/docs/latest/use/configure/combine-configs
+
+## Drawbacks
+
+1. Introducing a new key to flat config at this point means that users of ESLint v8 won't be able to use it. Even though v8 is EOL, we had made an attempt to provide forward compatibility to make people's transition easier.
+1. While this may reduce complexity for users, it increases the complexity of config evaluation inside of ESLint. This necessarily means a performance cost that we won't be able to quantify until implementation.
+1. Users will have to know which version of ESLint supports `extends` in order to use it. There really isn't an easy way to feature test this capability other than to inspect the version of the current ESLint package.
+
+## Backwards Compatibility Analysis
+
+This proposal is additive and does not affect the way existing configurations are evaluated.
+
+## Alternatives
+
+
+### A `defineConfig()` function
+
+One alternative is to create a `defineConfig()` function that is exported from the `eslint` package. All of the functionality described in this RFC could be implemented in that function, leaving both `ConfigArray` and `FlatConfigArray` in their current state without the need for changes.
+
+```js
+import globals from "globals";
+import { defineConfig } from "eslint";
+
+const config1 = {
+    name: "config1",
+    files: ["**/*.cjs.js"],
+    languageOptions: {
+        sourceType: "commonjs",
+        globals: {
+            ...globals.node
+        }
+    }
+};
+
+const config2 = {
+    name: "config2",
+    files: ["**/*.js"],
+    ignores: ["**/tests"],
+    rules: {
+        "no-console": "error"
+    }
+};
+
+
+export default defineConfig([
```

Comment:
Given we're also proposing calling `.flat(Infinity)` -- I would suggest making this a variadic function.

This would allow the flexibility of doing both: `defineConfig([ {...}, {...} ])` and `defineConfig({...}, {...})` and having the same config as a result.

It's a nice thing to be able to drop the explicit extra `[]` wrapping inside of the function parens `()`.

Reply:
Fair point. 👍